src/H5Odtype.c

```diff
@@ -563,14 +752,6 @@ H5O_dtype_decode_helper(H5F_t *f, unsigned *ioflags /*in,out*/, const uint8_t **
     } /* end switch */
 
 done:
-    if (ret_value < 0) {
-        if (dt != NULL) {
-            if (dt->shared != NULL)
-                dt->shared = H5FL_FREE(H5T_shared_t, dt->shared);
-            dt = H5FL_FREE(H5T_t, dt);
-        } /* end if */
-    }     /* end if */
```

Comment:
This should match develop in that develop has:

```C
    /* Cleanup on error */
    if (ret_value < 0)
        /* Release (reset) dt but do not free it - leave it as an empty datatype as was the case on
         * function entry */
        if (H5T__free(dt) < 0)
            HDONE_ERROR(H5E_DATATYPE, H5E_CANTRELEASE, FAIL, "can't release datatype info")
```

here to fix a separate issue.